server/decoders/utils.ts

```diff
@@ -36,7 +54,13 @@ export function jsonArrayString<T>(itemDecoder : Decoder<T>) : Decoder<Array<T>>
 
 export function jsonObjectString<T>(objectDecoder : Decoder<T>) : Decoder<T>
 {
-    const decodeJsonStr = map(string, (str) => JSON.parse(str));
+    const decodeJsonStr = map(
+        either(string, inexact({})),
+        (item) =>
+        {
+            return typeof item === 'string' ? JSON.parse(item) : item;
+        }
+    );
```

Comment:
These seem a bit off. Seems like it’s trying to handle an already decoded object. This seems like a problem better solved at the call site, which the type system should hopefully help with.

Reply:
Not entirely. To put this decoder into english:

> "Take the value and decode it as either a string or an object. If it's a string, parse it as json. Now that we're sure we have an object, apply `objectDecoder`."

Normally, you're right, I'd want to be consistent about the inputs. But the problem here is that the callsite is generic. It's the same code for all possible supplements. So it's really hard to have this one supplement, in this one system do a `JSON.parse` on a key before sending it in to the decode. (This decoder originally just handled string values that were really objects, all I added was support for if it was already a object, skip the string parsing.)

The most elegant way to solve this? Probably not. But I'm struggling to find a solution that's cleaner.

Reply:
Also, I blame SQLite for not supporting parsing of their json fields automatically.

Reply:
I guess I just lack full enough context here.